### PR TITLE
Support Preflight CORS requests

### LINKED ISSUE #44: Middlewares aren't called if there's no matched route
I found that this was also discussed in the old repository in https://github.com/ZijianHe/koa-router/issues/182 https://github.com/ZijianHe/koa-router/issues/257 https://github.com/ZijianHe/koa-router/issues/239.

The problem can be seen with a code like this. The testcase could be made simpler but I left it like this so that it's closer to my real-world case, implementing CORS.

```js
// @flow
const Koa = require('koa');
const Router = require('@koa/router');
const cors = require('@koa/cors');

function getRouter() {
  const router = new Router();

  router.get('/hello1', ctx => {
    ctx.body = 'world1';
  });
  router.use(cors());
  router.post('/hello2', ctx => {
    ctx.body = 'world2';
  });

  return router;
}

const app = new Koa();
const router = getRouter();
app.use(router.routes());
app.use(router.allowedMethods());

app.listen(20000);
console.log('server started on port 20000');
```

This is a curl command you can use to test this (this is a socalled "preflight CORS request"):
```
curl -i -H 'Origin: foo' -H 'Access-Control-Request-Method: POST' -X OPTIONS http://localhost:20000/hello2
```

The answer should be something like:
```
HTTP/1.1 204 No Content
Vary: Origin
Access-Control-Allow-Origin: foo
Access-Control-Allow-Methods: GET,HEAD,PUT,POST,DELETE,PATCH
Date: Fri, 10 Jan 2020 18:36:33 GMT
Connection: keep-alive
```
But instead it is:
```
HTTP/1.1 200 OK
Content-Type: text/plain; charset=utf-8
Content-Length: 0
Allow: POST
Date: Fri, 10 Jan 2020 18:36:28 GMT
Connection: keep-alive
```
This shows that this is the middleware of `allowedMethods` that runs, instead of the "cors" middleware.

Note this does work:
```js
// @flow
const Koa = require('koa');
const Router = require('@koa/router');
const cors = require('@koa/cors');

function getRouter1() {
  const router = new Router();

  router.get('/hello1', ctx => {
    ctx.body = 'world1';
  });

  return router;
}

function getRouter2() {
  const router = new Router();

  router.post('/hello2', ctx => {
    ctx.body = 'world2';
  });

  return router;
}

const app = new Koa();
const router1 = getRouter1();
app.use(router1.routes());
app.use(router1.allowedMethods());

app.use(cors());

const router2 = getRouter2();
app.use(router2.routes());
app.use(router2.allowedMethods());

app.listen(20001);
console.log('server started on port 20001');
```

My main issue is that there's a surprise here: the router doesn't behave like koa and ignores all middleware if there's no matching route, and it's not documented anywhere. Koa instead will always run all middleware.

After various tries, I believe that this simple patch makes it work properly and doesn't have much impact:
```diff
diff --git a/lib/router.js b/lib/router.js
index a97941f..e681b6f 100644
--- a/lib/router.js
+++ b/lib/router.js
@@ -331,23 +331,24 @@ Router.prototype.routes = Router.prototype.middleware = function () {
     if (ctx.matched) {
       ctx.matched.push.apply(ctx.matched, matched.path);
     } else {
       ctx.matched = matched.path;
     }
 
     ctx.router = router;
 
-    if (!matched.route) return next();
-
     var matchedLayers = matched.pathAndMethod
-    var mostSpecificLayer = matchedLayers[matchedLayers.length - 1]
-    ctx._matchedRoute = mostSpecificLayer.path;
-    if (mostSpecificLayer.name) {
-      ctx._matchedRouteName = mostSpecificLayer.name;
+
+    if (matched.route) {
+      var mostSpecificLayer = matchedLayers[matchedLayers.length - 1]
+      ctx._matchedRoute = mostSpecificLayer.path;
+      if (mostSpecificLayer.name) {
+        ctx._matchedRouteName = mostSpecificLayer.name;
+      }
     }
 
     layerChain = matchedLayers.reduce(function(memo, layer) {
       memo.push(function(ctx, next) {
         ctx.captures = layer.captures(path, ctx.captures);
         ctx.params = layer.params(path, ctx.captures, ctx.params);
         ctx.routerName = layer.name;
         return next();
```

Please tell me if you'd like that I do a PR with some test changes, I can do it.

### PR DESCRIPTION
**Pull request by julienw**
Fixes #44

Kindly see issue #44 for more information about this patch. Thanks!

### CODE REVIEW

**Comment by JacobMGEvans** [comment]
@niftylettuce thoughts?

**Comment by niftylettuce** [comment]
Seems like this will require a major version bump, would you agree @JacobMGEvans 	?

**Comment by JacobMGEvans** [comment]
I am not seeing any breaking changes, and issue #44 details out the change really well (probably needs to be added to some documentation).

However, people might have code set up expecting the 200 behavior therefore I would agree that a major bump with a release note annotating the change would be a good idea.

**Comment by moritzruth** [comment]
When are you planning to release it?

**Comment by Sinewyk** [comment]
I added info in #44 

TL;DR: it's not a bug, it's a feature.
